File: Lab1/cs_solution/TextData.cs
using System.Text.RegularExpressions;

record TextData(
    string fileName,
    string Text,
    int nOfVowels,
    int nOfConsonants,
    int nOfLetters,
    int nOfSentences,
    string longestWord
    )
{

    public static TextData ProcessText(string fileName, string text)
    {
        int vows = CountVows(text);
        int cons = CountCons(text);
        int letters = vows + cons;
        int nsent = CountSentences(text);
        string? longestWord = FindLongWord(text);
        return new TextData(fileName, text, vows, cons, letters, nsent, longestWord ?? "");
    }

    private static int CountVows(string text) =>
        text.Count(val => "aeuiohAEUIOH".Contains(val));

    private static int CountCons(string text) =>
        text.Count(val => char.IsAsciiLetter(val) && !"aeuiohAEUIOH".Contains(val));

    private static int CountSentences(string text) =>
        Regex.Matches(text, @"[!?.](/s | $)").Count;
    
    private static string? FindLongWord(string text){
        MatchCollection matches = Regex.Matches(text, @"\w+");
        return matches
            .Cast<Match>()
            .OrderByDescending(m => m.Length)
            .FirstOrDefault()?.Value;
    } 

}

File: Lab1/ts_solution/src/modules/TextData.ts
export class TextData{
    constructor(
        public fileName: string,
        public text: string,
        public nOfVowels:number,
        public nOfCons:number,
        public nOfLetters:number,       
        public nOfSentences: number,
        public longestWord: string,
    ){}
    static ProcessText(fileName: string, text: string): TextData{
        var vows = this.CountVows(text);
        var cons = this.CountCons(text);
        var letters = vows + cons;
        var nSent = this.CountSentences(text);
        var longestWord = this.FindLongWord(text);
        return new TextData(fileName, text, vows, cons, letters, nSent, longestWord || "")
    }
    static CountVows(text: string):number{
        var vowels = "aeuiohAEUIOH"//["a", "e", "u", "i", "o", "h", "A", "E", "U", "I", "O", "H"]
        return Array.from(text).filter(char => vowels.includes(char)).length;
    }
    
    static CountCons(text: string):number{
        var vowels = ["a", "e", "u", "i", "o", "h", "A", "E", "U", "I", "O", "H"]
        return Array.from(text).filter(char => !vowels.includes(char)).length;
    }
    static CountSentences(text: string):number{
        return (text.match(/[!?.]/g) || []).length;
    }
    static FindLongWord(text:string):string | null {
        const matches = text.match(/\w+/g);

        if (!matches) return null;

        return matches.reduce((longest, current) => 
            current.length > longest.length ? current : longest, "");
    }
    
}
